### PR TITLE
Improvement for github action for building website

### PR DESCRIPTION
**Pull request by k1ng440**
This pull request is to improve the speed of the workflow

**Changes:** 
- New builder docker images 
- Added caching for  emscripten
- Added caching for yarn
- Added `-DBUILD_TESTING=NO` to C++ configure
- Replaced `cd website` with `working-directory`  in a couple of steps to make it more readable

### CODE REVIEW

File: .github/docker/quick-lint-js-github-website-builder/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
+FROM ubuntu:20.04
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN \
+    apt-get update && \
+    apt-get install -y --no-install-recommends \
```
Comment (strager) [inline .github/docker/quick-lint-js-github-website-builder/Dockerfile:7]: Must fix: Put `apt-get update` in a separate `RUN` from `apt-get install` like in our other `Dockerfile`s. (I know it might be against "best practice".)

File: .github/docker/quick-lint-js-github-website-builder/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    apt-add-repository 'deb https://deb.nodesource.com/node_14.x focal main' && \
+    sed -n -e 's/^deb /deb-src /p' /etc/apt/sources.list >/etc/apt/sources.list.d/quick-lint-js-sources.list && \
+    apt-get install -y --no-install-recommends nodejs && \
+    npm install -g yarn && \
+    # install python packages
+    pip install bs4 requests && \
+    # configure apache2
+    a2enmod deflate headers && \
+    sed -i -e 's/AllowOverride\ None/AllowOverride\ FileInfo/' /etc/apache2/apache2.conf && \
+    rm -rf /var/lib/apt/lists/*
```
Comment (strager) [inline .github/docker/quick-lint-js-github-website-builder/Dockerfile:33]: Put this in a separate `RUN`.

File: .github/workflows/build-website.yml [hunk 3]
```diff
@@ -1,75 +1,91 @@
-# Copyright (C) 2020  Matthew Glazar
-# See end of file for extended copyright information.
```
Comment (strager) [inline .github/workflows/build-website.yml:2]: Must fix: Keep this comment.

File: .github/workflows/build-website.yml [hunk 4]
```diff
@@ -1,75 +1,91 @@
...
+        id: cache-system-libraries
+        uses: actions/cache@v2
+        with:
+          path: ${{env.EM_CACHE_FOLDER}}
+          key: ${{env.EM_VERSION}}-${{ runner.os }}
+
+      - uses: quick-lint/setup-emsdk@v7
+        with:
+          version: ${{env.EM_VERSION}}
+          actions-cache-folder: ${{env.EM_CACHE_FOLDER}}
```
Comment (strager) [inline .github/workflows/build-website.yml:32]: What does this have to do with switching to a Docker container?

File: .github/workflows/build-website.yml [hunk 5]
```diff
@@ -1,75 +1,91 @@
...
+      - name: get yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+      - uses: actions/cache@v1
+        id: yarn-cache
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
```
Comment (strager) [inline .github/workflows/build-website.yml:50]: What does this have to do with switching to a Docker container?

File: .github/workflows/build-website.yml [hunk 6]
```diff
@@ -1,75 +1,91 @@
...
 
       - name: build site
-        run: cd website && yarn build
+        working-directory: website
+        run: yarn build
+
       - name: bundle site
-        run: cd website && zip -r ../website.zip www
+        working-directory: website
+        run: zip -r ../website.zip www
```
Comment (strager) [inline .github/workflows/build-website.yml:66]: Why these changes?

File: .github/workflows/build-website.yml [hunk 7]
```diff
@@ -1,75 +1,91 @@
...
-          ENV/bin/pip install bs4 requests
-      - name: check links (dev)
-        run: |
-          cd website/
-          yarn start localhost 9001 &
-          sleep 0.5  # HACK(strager): Wait for server to start.
-          cd ../
-          ./ENV/bin/python ./tools/check-links http://localhost:9001/
-          kill %1
-        shell: bash
```
Comment (strager) [inline .github/workflows/build-website.yml:55]: Must fix: Why did you delete `check links (dev)`? This step is important to make sure the dev server works. Undelete this step.

File: .github/workflows/build-website.yml [hunk 8]
```diff
@@ -1,75 +1,91 @@
...
+ 
+      - name: Upload build to workflow
         uses: actions/upload-artifact@v2
         with:
           if-no-files-found: error
           name: website-${{ github.sha }}
           path: website/www/
-      - name: upload build to long-term storage
+
+      - name: deploy website
```
Comment (strager) [inline .github/workflows/build-website.yml:88]: Must fix: This change is wrong. The last step does *not* deploy anything. It copies files to an artifact server. The `website/deploy.sh` script deploys. Undo this change.

**Review by strager — CHANGES_REQUESTED** [review]
This patch seems to do a few unrelated things. Can you split the things into separate commits or PRs, or enumerate the changes in the PR message?

File: .github/docker/quick-lint-js-github-website-builder/Dockerfile (hunk 1 above)
Comment (k1ng440) [reply .github/docker/quick-lint-js-github-website-builder/Dockerfile:7]: it will increase the size of the image though

File: .github/workflows/build-website.yml (hunk 4 above)
Comment (k1ng440) [reply .github/workflows/build-website.yml:32]: this pull request is a kind of complete overhaul of this workflow.

File: .github/workflows/build-website.yml (hunk 5 above)
Comment (k1ng440) [reply .github/workflows/build-website.yml:50]: this is to speed up the workflow.

File: .github/workflows/build-website.yml (hunk 6 above)
Comment (k1ng440) [reply .github/workflows/build-website.yml:66]: because I think leveraging the `working-directory` is cleaner than doing `cd website`

File: .github/workflows/build-website.yml (hunk 8 above)
Comment (k1ng440) [reply .github/workflows/build-website.yml:88]: sure.

File: .github/workflows/build-website.yml (hunk 7 above)
Comment (k1ng440) [reply .github/workflows/build-website.yml:55]: ok, why do we need to check links using apache2 then?

Comment (strager) [reply .github/workflows/build-website.yml:55]: Because the dev server and the build can have different characteristics, since the code isn't 100% identical between the two. We've had bugs in each one.

File: .github/workflows/build-website.yml [hunk 9]
```diff
@@ -11,64 +11,92 @@ jobs:
...
+        id: cache-system-libraries
+        uses: actions/cache@v2
+        with:
+          path: ${{env.EM_CACHE_FOLDER}}
+          key: ${{env.EM_VERSION}}-${{ runner.os }}
+
+      - uses: quick-lint/setup-emsdk@v7
+        with:
+          version: ${{env.EM_VERSION}}
+          actions-cache-folder: ${{env.EM_CACHE_FOLDER}}
```
Comment (strager) [inline .github/workflows/build-website.yml:34]: We should really put the emscripten installation in the Docker container instead.

File: .github/workflows/build-website.yml [hunk 10]
```diff
@@ -11,64 +11,92 @@ jobs:
...
           yarn start localhost 9001 &
           sleep 0.5  # HACK(strager): Wait for server to start.
           cd ../
-          ./ENV/bin/python ./tools/check-links http://localhost:9001/
+          python3 ./tools/check-links http://localhost:9001/
           kill %1
         shell: bash
-      - name: check links (apache2)
+
+      - name: check links
```
Comment (strager) [inline .github/workflows/build-website.yml:80]: Keep the old name.
```suggestion
      - name: check links (apache2)
```

File: .github/workflows/build-website.yml [hunk 11]
```diff
@@ -2,7 +2,7 @@
 # See end of file for extended copyright information.
 
 name: build website
-on:
+on:
```
Comment (strager) [inline .github/workflows/build-website.yml:5]: Nit: Undo trivial whitespace changes like this.

File: .github/workflows/build-website.yml [hunk 12]
```diff
@@ -11,64 +11,92 @@ jobs:
...
+      - name: get yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+      - uses: actions/cache@v1
+        id: yarn-cache
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
```
Comment (strager) [inline .github/workflows/build-website.yml:52]: How does `restore-keys` work in this case? From what I can tell, this will basically accomplish nothing. If that's true, then why specify `restore-keys` at all?

File: .github/workflows/build-website.yml [hunk 13]
```diff
@@ -11,64 +11,92 @@ jobs:
...
         run: emmake cmake --install build --component vscode --prefix website/public/demo
+      
+      - name: get yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+      - uses: actions/cache@v1
+        id: yarn-cache
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
```
Comment (strager) [inline .github/workflows/build-website.yml:50]: We only care about `website/yarn.lock`, not the other lock files, right?

**Review by strager — APPROVED** [review]

File: .github/workflows/build-website.yml (hunk 13 above)
Comment (k1ng440) [reply .github/workflows/build-website.yml:50]: yes. that is correct

File: .github/workflows/build-website.yml (hunk 9 above)
Comment (k1ng440) [reply .github/workflows/build-website.yml:34]: size of emscripten is about 1.11 GB. see [screenshot](https://puu.sh/Jd5rf/7794b36cde.png)

File: .github/docker/quick-lint-js-github-website-builder/README.md [hunk 14]
```diff
@@ -0,0 +1,13 @@
+# quick-lint-js-website-builder Docker image
+
+## Updating the Docker image
+
+Pick a version number for the new Docker image (e.g. `v3`), then run the
+following commands:
+
+    $ docker build --tag ghcr.io/quick-lint/quick-lint-js-website-builder:VERSION_NUMBER_HERE .github/docker/quick-lint-js-website-builder/
```
Comment (strager) [inline .github/docker/quick-lint-js-github-website-builder/README.md:8]: The name of the image is incorrect (missing `github-`). I'll fix it when I merge.

**Comment by strager** [comment]
Landed as commit a4e1437ec3d0f615fba612c0e0f8b3afa87cc780. Thanks!
